packages/@aws-cdk/aws-codepipeline/lib/pipeline.ts

```diff
@@ -459,21 +468,55 @@ export class Pipeline extends PipelineBase {
 
     // check whether we already have a bucket in that region,
     // either passed from the outside or previously created
+    const crossRegionSupport = this.obtainCrossRegionSupportFor(action, actionEnv.region);
+
+    // the stack containing the replication bucket must be deployed before the pipeline
+    Stack.of(this).addDependency(crossRegionSupport.stack);
+    // The Pipeline role must be able to replicate to that bucket
+    crossRegionSupport.replicationBucket.grantReadWrite(this.role);
+
+    return {
+      artifactBucket: crossRegionSupport.replicationBucket,
+      region: actionEnv.region,
+    };
+  }
+
+  /**
+   * Get or create the cross-region support construct for the given action
+   */
+  private obtainCrossRegionSupportFor(action: IAction, actionRegion: string) {
     let crossRegionSupport = this._crossRegionSupport[actionRegion];
     if (!crossRegionSupport) {
       // we need to create scaffolding resources for this region
+      const otherStack = this.crossRegionSupportScope(action);
       crossRegionSupport = this.createSupportResourcesForRegion(otherStack, actionRegion);
       this._crossRegionSupport[actionRegion] = crossRegionSupport;
     }
+    return crossRegionSupport;
+  }
 
-    // the stack containing the replication bucket must be deployed before the pipeline
-    pipelineStack.addDependency(crossRegionSupport.stack);
-    crossRegionSupport.replicationBucket.grantReadWrite(this.role);
+  /**
+   * Return the correct scope to create the cross-region construct in
+   *
+   * As an optimization, we reuse the action resource's stack if the stack has
+   * the same environment as the action.
+   *
+   * (Always true for owned action resources but not necessarily true
+   * for imported ones).
+   */
+  private crossRegionSupportScope(action: IAction): Stack | undefined {
+    // The resource this action is targeting (if any). Might be owned, might be imported.
+    const actionResource = action.actionProperties.resource;
+    if (!actionResource) { return undefined; }
 
-    return {
-      artifactBucket: crossRegionSupport.replicationBucket,
-      region: actionRegion,
-    };
+    const actionResourceStack = Stack.of(actionResource);
+
+    const stackHasCorrectEnvironment = (
+      sameEnvDimension(actionResource.env.region, actionResourceStack.region)
+      && sameEnvDimension(actionResource.env.account, actionResourceStack.account));
```

Comment:
1. I think the parens are redundant...?

```suggestion
    const stackHasCorrectEnvironment =
      sameEnvDimension(actionResource.env.region, actionResourceStack.region) &&
      sameEnvDimension(actionResource.env.account, actionResourceStack.account);
```

2. I don't really like the name `stackHasCorrectEnvironment`, it doesn't actually convey what this variable represents in my mind. How about `actionResourceIsFromSameEnvAsItsOwningStack`? Or something like that 🙂.

Reply:
1. I think I put those parens there because I thought the linter would yell at me otherwise.

2. Geez we are really not compatible in how we like our source, huh? :) 

I prefer my logical combinators at the start of the line, so I can easily scan how the clauses are combined. Is it a `&&` or a `||`? At the end of the line you have to hunt for them.

`actionResourceIsFromSameEnvAsItsOwningStack` -- this just hurts my eyes and my comprehension. I prefer short terms (for readability/scannability) and abstractions. It's enough to know that something is the "correct" environment. I can now make an abstraction in my head saying that something can be the "correct" or "not the correct" environment, and then some decisions are based off of that.

If I'm curious about what makes an environment "correct", I can always look for its definition.

In other words, "correct" is the conclusion for which `IsFromSameEnv` is the question. I prefer not to keep asking the question on every line.